File: c/grains.c
#include "grains.h"

unsigned long long square(int n) {
  if (n < 1 || n > 64)
    return 0ul;

  unsigned long long res = 1;
  int base = 2;

  n--;
  while (n > 0) {
    res *= base;
    n--;
  }
  return res;
}
unsigned long long total() {
  unsigned long long tot = 0ul;
  for (int i = 1; i <= 64; i++) {
    tot += square(i);
  }
  return tot;
}

File: go/grains.go
package grains

import (
	"errors"
	"math"
)

// Square calculate how many grains on the nth square
func Square(n int) (uint64, error) {
	if n <= 0 || n > 64 {
		return 0, errors.New("n must bettwen 0 and 64")
	}

	return uint64(math.Pow(2, float64(n-1))), nil
}

// Total number of grains
func Total() (tot uint64) {
	for i := 1; i <= 64; i++ {
		s, _ := Square(i)
		tot += s
	}
	return
}
